File: plex_cli/menus/media.py
"""Media item sub-menu.

Provides actions for a selected Plex media item.
"""

from __future__ import annotations

from typing import Any

import questionary
from rich.console import Console

from plex_cli.api import PlexAPIError, PlexClient
from plex_cli.display import (
    console,
    error,
    header,
    info,
    media_table,
    metadata_panel,
    success,
)

_MEDIA_ACTIONS = [
    questionary.Choice("View metadata", value="meta"),
    questionary.Choice("Browse children (seasons / episodes / tracks)", value="children"),
    questionary.Choice("Play on a client", value="play"),
    questionary.Choice("Mark as watched", value="watched"),
    questionary.Choice("Mark as unwatched", value="unwatched"),
    questionary.Choice("Add to playlist", value="playlist"),
    questionary.Choice("Show similar / related", value="similar"),
    questionary.Choice("Show extras (trailers, featurettes)", value="extras"),
    questionary.Separator(),
    questionary.Choice("← Back", value="back"),
]


def _choose_client(client: PlexClient) -> dict[str, Any] | None:
    """Let the user pick an available Plex client; return it or None."""
    try:
        clients = client.clients()
    except PlexAPIError as exc:
        error(str(exc))
        return None
    if not clients:
        error("No clients found.  Make sure a Plex player is open.")
        return None
    choices = [questionary.Choice(c.get("name", "?"), value=c) for c in clients]
    choices.append(questionary.Choice("← Cancel", value=None))
    return questionary.select("Choose a client:", choices=choices).ask()


def run_media_menu(client: PlexClient, item: dict[str, Any]) -> None:
    """Interactive sub-menu for a single media item."""
    title = item.get("title", "?")
    rating_key = item.get("ratingKey", "")

    while True:
        header(f"Media: {title}")
        action = questionary.select(
            "What would you like to do?",
            choices=_MEDIA_ACTIONS,
        ).ask()

        if action is None or action == "back":
            break

        if action == "meta":
            try:
                full = client.metadata(rating_key)
            except PlexAPIError as exc:
                error(str(exc))
                continue
            console.print(metadata_panel(full))
            questionary.press_any_key_to_continue("Press any key to continue…").ask()

        elif action == "children":
            try:
                kids = client.children(rating_key)
            except PlexAPIError as exc:
                error(str(exc))
                continue
            if not kids:
                info("No children found.")
                continue
            console.print(media_table(kids, title=f"Children of {title}"))
            chosen = questionary.select(
                "Select an item (or back):",
                choices=[questionary.Choice(k.get("title", "?"), value=k) for k in kids]
                + [questionary.Choice("← Back", value=None)],
            ).ask()
            if chosen:
                run_media_menu(client, chosen)

        elif action == "play":
            chosen_client = _choose_client(client)
            if not chosen_client:
                continue
            address = chosen_client.get("address", "")
            port = str(chosen_client.get("port", "32433"))
            try:
                client.play_media(address, port, rating_key)
                success(f"Sent play command to {chosen_client.get('name', '?')}")
            except PlexAPIError as exc:
                error(str(exc))

        elif action == "watched":
            try:
                client.mark_watched(rating_key)
                success("Marked as watched.")
            except PlexAPIError as exc:
                error(str(exc))

        elif action == "unwatched":
            try:
                client.mark_unwatched(rating_key)
                success("Marked as unwatched.")
            except PlexAPIError as exc:
                error(str(exc))

        elif action == "playlist":
            try:
                playlists = client.playlists()
            except PlexAPIError as exc:
                error(str(exc))
                continue
            if not playlists:
                info("No playlists found.")
                continue
            choices = [questionary.Choice(p.get("title", "?"), value=p) for p in playlists]
            choices.append(questionary.Choice("← Cancel", value=None))
            chosen_pl = questionary.select("Add to which playlist?", choices=choices).ask()
            if not chosen_pl:
                continue
            pl_id = chosen_pl.get("ratingKey", "")
            try:
                client.add_to_playlist(pl_id, rating_key)
                success(f"Added to '{chosen_pl.get('title')}'.")
            except PlexAPIError as exc:
                error(str(exc))

        elif action == "similar":
            try:
                similar = client.similar(rating_key)
            except PlexAPIError as exc:
                error(str(exc))
                continue
            if not similar:
                info("No similar items found.")
                continue
            console.print(media_table(similar, title="Similar"))
            questionary.press_any_key_to_continue("Press any key…").ask()

        elif action == "extras":
            try:
                extras = client.extras(rating_key)
            except PlexAPIError as exc:
                error(str(exc))
                continue
            if not extras:
                info("No extras found.")
                continue
            console.print(media_table(extras, title="Extras"))
            questionary.press_any_key_to_continue("Press any key…").ask()

File: src/menus/media.ts
/**
 * Media item sub-menu.
 */

import inquirer from 'inquirer';
import {
  metadataPanel,
  mediaTable,
  printError,
  printSuccess,
  printInfo,
  printHeader,
} from '../display';
import type { PlexClient } from '../api';
import type { PlexItem, PlexClientDevice } from '../types';

async function chooseClient(client: PlexClient): Promise<PlexClientDevice | null> {
  let clients: PlexClientDevice[];
  try {
    clients = await client.clients();
  } catch (err: unknown) {
    printError((err as Error).message);
    return null;
  }
  if (!clients.length) {
    printError('No clients found. Make sure a Plex player is open.');
    return null;
  }
  const { chosen } = await inquirer.prompt<{ chosen: PlexClientDevice | null }>([
    {
      type: 'list',
      name: 'chosen',
      message: 'Choose a client:',
      choices: [
        ...clients.map((c) => ({ name: c.name ?? c.title ?? '?', value: c })),
        { name: '← Cancel', value: null },
      ],
    },
  ]);
  return chosen;
}

export async function runMediaMenu(client: PlexClient, item: PlexItem): Promise<void> {
  const title = item.title ?? '?';
  const ratingKey = item.ratingKey ?? '';

  while (true) {
    printHeader(`Media: ${title}`);
    const { action } = await inquirer.prompt<{ action: string }>([
      {
        type: 'list',
        name: 'action',
        message: 'What would you like to do?',
        choices: [
          { name: 'View metadata', value: 'meta' },
          { name: 'Browse children (seasons / episodes / tracks)', value: 'children' },
          { name: 'Play on a client', value: 'play' },
          { name: 'Mark as watched', value: 'watched' },
          { name: 'Mark as unwatched', value: 'unwatched' },
          { name: 'Add to playlist', value: 'playlist' },
          { name: 'Show similar / related', value: 'similar' },
          { name: 'Show extras (trailers, featurettes)', value: 'extras' },
          new inquirer.Separator(),
          { name: '← Back', value: 'back' },
        ],
      },
    ]);

    if (action === 'back') break;

    if (action === 'meta') {
      try {
        const full = await client.metadata(ratingKey);
        metadataPanel(full);
      } catch (err: unknown) {
        printError((err as Error).message);
      }
      await pressAnyKey();

    } else if (action === 'children') {
      try {
        const kids = await client.children(ratingKey);
        if (!kids.length) { printInfo('No children found.'); continue; }
        mediaTable(kids, `Children of ${title}`);
        const { chosen } = await inquirer.prompt<{ chosen: PlexItem | null }>([
          {
            type: 'list',
            name: 'chosen',
            message: 'Select an item (or back):',
            choices: [
              ...kids.map((k) => ({ name: k.title ?? '?', value: k })),
              { name: '← Back', value: null },
            ],
          },
        ]);
        if (chosen) await runMediaMenu(client, chosen);
      } catch (err: unknown) {
        printError((err as Error).message);
      }

    } else if (action === 'play') {
      const chosenClient = await chooseClient(client);
      if (!chosenClient) continue;
      try {
        await client.playMedia(
          chosenClient.address ?? '',
          String(chosenClient.port ?? '32433'),
          ratingKey
        );
        printSuccess(`Sent play command to ${chosenClient.name ?? '?'}`);
      } catch (err: unknown) {
        printError((err as Error).message);
      }

    } else if (action === 'watched') {
      try {
        await client.markWatched(ratingKey);
        printSuccess('Marked as watched.');
      } catch (err: unknown) {
        printError((err as Error).message);
      }

    } else if (action === 'unwatched') {
      try {
        await client.markUnwatched(ratingKey);
        printSuccess('Marked as unwatched.');
      } catch (err: unknown) {
        printError((err as Error).message);
      }

    } else if (action === 'playlist') {
      try {
        const playlists = await client.playlists();
        if (!playlists.length) { printInfo('No playlists found.'); continue; }
        const { chosen } = await inquirer.prompt<{ chosen: PlexItem | null }>([
          {
            type: 'list',
            name: 'chosen',
            message: 'Add to which playlist?',
            choices: [
              ...playlists.map((p) => ({ name: p.title ?? '?', value: p })),
              { name: '← Cancel', value: null },
            ],
          },
        ]);
        if (!chosen) continue;
        await client.addToPlaylist(chosen.ratingKey ?? '', ratingKey);
        printSuccess(`Added to '${chosen.title}'.`);
      } catch (err: unknown) {
        printError((err as Error).message);
      }

    } else if (action === 'similar') {
      try {
        const similar = await client.similar(ratingKey);
        if (!similar.length) { printInfo('No similar items found.'); continue; }
        mediaTable(similar, 'Similar');
      } catch (err: unknown) {
        printError((err as Error).message);
      }
      await pressAnyKey();

    } else if (action === 'extras') {
      try {
        const extras = await client.extras(ratingKey);
        if (!extras.length) { printInfo('No extras found.'); continue; }
        mediaTable(extras, 'Extras');
      } catch (err: unknown) {
        printError((err as Error).message);
      }
      await pressAnyKey();
    }
  }
}

async function pressAnyKey(msg = 'Press Enter to continue…'): Promise<void> {
  await inquirer.prompt([{ type: 'input', name: '_', message: msg }]);
}
